### PR TITLE
Use only `FoundationEssentials` when possible

### PR DESCRIPTION
**Pull request by t089**
Linking against `Foundation` on linux adds a hefty binary size penalty. Mostly due to the ICU data from `FoundationInternationalization`. To address this we conditionally import `FoundationEssentials` instead of `Foundation` if possible and replace `Foundation`-only API with a different implementation.

### Motivation:

Smaller binary sizes are better.

### Modifications:

- Replace `import Foundation` with `import FoundationEssentials`.
- Update date parsing logic to use newer API if available.
- Implement a custom parser for RFC5322 dates

### Result:

Ability to build the package without linking `Foundation`.

### CODE REVIEW

File: Sources/AWSLambdaEvents/Utils/DateWrappers.swift [hunk 1]
```diff
@@ -84,34 +113,256 @@ public struct RFC5322DateTimeCoding: Decodable, Sendable {
...
-        for formatter in Self.dateFormatters {
-            if let date = formatter.date(from: string) {
-                self.wrappedValue = date
-                return
+    }
+}
+
+struct RFC5322DateParsingError: Error {}
+
+struct RFC5322DateParseStrategy {
```
Comment (0xTim) [inline Sources/AWSLambdaEvents/Utils/DateWrappers.swift:136]: Do we have any tests for this? This is a lot of untested code, and a lot of unsafe code that we should really ensure works properly if there are not other alternatives

Comment (t089) [reply Sources/AWSLambdaEvents/Utils/DateWrappers.swift:136]: Yeah that's fair, it is currently being tested indirectly through the existing tests that parse dates.

Comment (t089) [reply Sources/AWSLambdaEvents/Utils/DateWrappers.swift:136]: We can add more test for some of edge cases. I am also not sure if we need to support more than the handful of hard-coded named timezones I added. My assumption is though that S3 will always retuned well-formed dates in UTC.

Comment (t089) [reply Sources/AWSLambdaEvents/Utils/DateWrappers.swift:136]: A regarding "unsafe". I think all API is actually safe. While we are operating on an "Unsafe"BufferPointer, we are only using its iterator methods, which should guarantee that there is no way to escape the buffer bounds by doing so.

Comment (adam-fowler) [reply Sources/AWSLambdaEvents/Utils/DateWrappers.swift:136]: I did suggest to the Foundation team a while back to include an HTTP datetime FormatStyle in FoundationEssentials and they seemed open to the idea. I never chased it up though. @t089 would you be open to pitching it with a version of this code?

**Review by sebsto — APPROVED** [review]
LGTM

**Comment by sebsto** [comment]
Thank you @t089 !
There is an issue with the license headers on some files and a couple of formatting checks that fails to validate 
See https://github.com/swift-server/swift-aws-lambda-events/actions/runs/12408957260 for the details 

Toc heck the formatting, I usually download the script from https://github.com/swiftlang/github-workflows/blob/main/.github/workflows/scripts/check-swift-format.sh and run it locally.

File: Sources/AWSLambdaEvents/Utils/DateWrappers.swift [hunk 2]
```diff
@@ -25,23 +29,44 @@ public struct ISO8601Coding: Decodable, Sendable {
...
                 debugDescription:
                     "Expected date to be in ISO8601 date format, but `\(dateString)` is not in the correct format"
             )
         }
-        self.wrappedValue = date
     }
 
+    private static func parseISO8601(dateString: String) throws -> Date {
+        if #available(macOS 12.0, *) {
+            return try Date(dateString, strategy: .iso8601)
```
[the remark this answers is not in the document]
Comment (t089) [reply Sources/AWSLambdaEvents/Utils/DateWrappers.swift:48]: Yeah there are a bunch of build errors with swift 5.10 on linux. will look into it.

[the remark this answers is not in the document]
Comment (sebsto) [reply Sources/AWSLambdaEvents/Utils/DateWrappers.swift:48]: The next version will support Swift 6 and onwards, aligned on the Swift Lambda Runtime

[the remark this answers is not in the document]
Comment (sebsto) [reply Sources/AWSLambdaEvents/Utils/DateWrappers.swift:48]: Let me know if you want to fix this or not. I'm happy to merge with a Swift 6 only change. If there is an easy fix, let's be nice with our existing users.

**Comment by sebsto** [comment]
@t089 looks like your last commit breaks the Swift 6 unit tests 
https://github.com/swift-server/swift-aws-lambda-events/actions/runs/12413722400/job/34657011160?pr=81

File: Sources/AWSLambdaEvents/Utils/DateWrappers.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (t089) [reply Sources/AWSLambdaEvents/Utils/DateWrappers.swift]: Actually the `if #available(macOS 12, *)` check was redundant as the conditional compile already excludes macOS altogether. Now it should compile for both on linux and macOS.

**Comment by t089** [comment]
@sebsto should be good now! got a bit confused with all the platform specific checks...

**Comment by sebsto** [comment]
Merged - thank you !

**Comment by adam-fowler** [comment]
Given this is being used with more than HTTP headers I think we need to extend the number of timezone identifiers. The list is missing standard ones listed in RFC PST, EST, MST, CST (funnily RFC only lists US zones). Also does the parser deal with additional characters at the end it is common to see email dates written as follows `Sun, 22 Sep 2024 18:46:31 +0100 (BST)` with the three character timezone at the end.

**Comment by sebsto** [comment]
@adam-fowler these are valid concerns indeed. I created https://github.com/swift-server/swift-aws-lambda-events/issues/85 to follow up

**Comment by t089** [comment]
> Given this is being used with more than HTTP headers I think we need to extend the number of timezone identifiers. The list is missing standard ones listed in RFC PST, EST, MST, CST (funnily RFC only lists US zones).

This is fair, I was assuming this is only used with AWS APIs where UTC/GMT should suffice.

> Also does the parser deal with additional characters at the en. It is common to see email dates written as follows `Sun, 22 Sep 2024 18:46:31 +0100 (BST)` with the three character timezone at the end. 

Yes I think this should be covered:

```swift
    func testWithLeadingDayName() throws {
        let input = "Fri, 26 Jun 2020 03:04:03 -0500 (CDT)"
        let date = try strategy.parse(input)
        XCTAssertEqual("2020-06-26 08:04:03 +0000", date.description)
    }
```

The Tz in brackets is ignored.

**Comment by adam-fowler** [comment]
> This is fair, I was assuming this is only used with AWS APIs where UTC/GMT should suffice

For SES you are decoding the email header which could come from any email provider.
